agroal-pool/src/main/java/io/agroal/pool/ConnectionPool.java

```diff
@@ -319,14 +331,22 @@ private ConnectionHandler handlerFromTransaction() throws SQLException {
 
     private ConnectionHandler handlerFromSharedCache() throws SQLException {
         long remaining = configuration.acquisitionTimeout().toNanos();
-        remaining = remaining > 0 ? remaining : Long.MAX_VALUE;
+        long deadline = remaining > 0 ? nanoTime() + remaining : MAX_VALUE;
         Future<ConnectionHandler> task = null;
         try {
-            for ( ; ; ) {
+            for ( int i = 0; ; i++ ) {
                 // If min-size increases, create a connection right away
                 if ( allConnections.size() < configuration.minSize() ) {
                     task = housekeepingExecutor.executeNow( new CreateConnectionTask() );
                 }
+                if ( i == 0 && handlerTransferQueue.hasWaitingConsumer() ) { // On the first iteration, block right away if there are other threads already blocked
+                    // There is a race condition here (if a thread was blocked but by the time poll is executed a transfer allready took place)
+                    // Because of that do not block for the whole remaining duration. Do it for at most a second and then move on to perform a scan
+                    ConnectionHandler handler = handlerTransferQueue.poll( Long.min( ONE_SECOND, remaining > 0 ? remaining * 9 / 10 : MAX_VALUE ), NANOSECONDS );
```

Comment:
you can put this out of this loop (and not in the i == 0 within the loop) and executing it regardless without checking the number of waiting consumers?

Reply:
it's not practical because of the code that comes before it on the loop